### PR TITLE
Shape Mouse refactor

### PR DESCRIPTION
**Pull request by Carreau**
Working on #2907 I realized the logic was a bit complex and could use
some generalisation, especially if we want to add more customisability.

This uses dictionaries and mapping to simply declare which callback
should be active with which modes, which should be easier to make
configurable, and also I believe easier to understand and less
error-prone (you only need to change what need to be active instead of
thinking about what need to be added and removed on mode change).

I've made the values class attributes, which mean that you should be
able to change the configuration at runtime by poking directly at the
class, and it should work as long as you don't change config of current
mode.

The only weird thing if maybe the _REV_SHAPE_HELP, I build the
dictionaries backward before of reversing it to only have a single
instance of each translated string as I'm not too sure how it deals with
duplication.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/layers/shapes/shapes.py [hunk 1]
```diff
@@ -297,6 +324,43 @@ class Shapes(Layer):
...
+    _drag_modes = {
+        Mode.SELECT: select,
+        Mode.DIRECT: select,
+        Mode.VERTEX_INSERT: vertex_insert,
+        Mode.VERTEX_REMOVE: vertex_remove,
+        Mode.ADD_RECTANGLE: add_rectangle,
+        Mode.ADD_ELLIPSE: add_ellipse,
+        Mode.ADD_LINE: add_line,
+        Mode.ADD_PATH: add_path_polygon,
+        Mode.ADD_POLYGON: add_path_polygon,
```
Comment (Carreau) [inline napari/layers/shapes/shapes.py:348]: I made all these privates as we could also have multiple callback per mode, but here it was not the case, 
so I use just callables as values.

File: napari/layers/shapes/shapes.py [hunk 2]
```diff
@@ -297,6 +324,43 @@ class Shapes(Layer):
...
+        Mode.SELECT: 'pointing',
+        Mode.DIRECT: 'pointing',
+        Mode.VERTEX_INSERT: 'cross',
+        Mode.VERTEX_REMOVE: 'cross',
+        Mode.ADD_RECTANGLE: 'cross',
+        Mode.ADD_ELLIPSE: 'cross',
+        Mode.ADD_LINE: 'cross',
+        Mode.ADD_PATH: 'cross',
+        Mode.ADD_POLYGON: 'cross',
+    }
```
Comment (Carreau) [inline napari/layers/shapes/shapes.py:374]: We could also add other checks, like that cursors are defined for all modes.

Comment (Carreau) [reply napari/layers/shapes/shapes.py:374]: I ended up doing this as some test were failing I introduced a no_op function and bind it when nothing should be done.

File: napari/layers/shapes/shapes.py [hunk 3]
```diff
@@ -58,6 +59,33 @@
...
+    trans._(
+        'hold <space> to pan/zoom, press <{BACKSPACE}> to remove selected',
+        BACKSPACE=BACKSPACE,
+    ): {Mode.SELECT, Mode.DIRECT},
+    trans._('enter a selection mode to edit shape properties'): {
+        Mode.PAN_ZOOM
+    },
+}
+
+_FWD_SHAPE_HELP = {}
```
Comment (sofroniewn) [inline napari/layers/shapes/shapes.py:88]: Can you add a comment here on why this was needed just so we don't forget. I could imagine coming back to this in a while and being confused

File: napari/layers/shapes/shapes.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/shapes/shapes.py:348]: private is good for me

File: napari/layers/shapes/shapes.py (hunk 2 above)
Comment (sofroniewn) [reply napari/layers/shapes/shapes.py:374]: sounds good

**Review by sofroniewn — APPROVED** [review]
Looks good to me @Carreau I like this change. Do you want to continue this for `Points`, `Labels` etc (all our layers with a mode) in follow up PRs. We could even eventually move some of the `mode` property/ setter to the base layer and then just let individual layers provide their own dictionaries - or some simple `ModeManager` / `Mode` object which defined the needed things. That can come later though. Maybe @jni wants to look before merge too though

File: napari/layers/shapes/shapes.py (hunk 3 above)
Comment (Carreau) [reply napari/layers/shapes/shapes.py:88]: Done.

**Comment by Carreau** [comment]
> Do you want to continue this for `Points`, `Labels` etc (all our layers with a mode) in follow up PRs. We could even eventually move some of the `mode` property/ setter to the base layer and then just let individual layers provide their own dictionaries - or some simple `ModeManager` / `Mode` object which defined the needed things.

Yes, that was the goal. 

Once it's in the base class I might even change to a dict of dict `mode_dict[mode][click|drag|double_click]`.

**Comment by sofroniewn** [comment]
> Once it's in the base class I might even change to a dict of dict mode_dict[mode][click|drag|double_click].

Ok great.

I'm merging now, as all this is internal refactoring of private stuff and more PRs are yet to come there is still time @jni if you'd like to see changes for them to come in the follow up PRs

**Pull request merged**